plugins/glustershd/actor.go

```diff
@@ -2,49 +2,66 @@ package glustershd
 
 import (
 	"errors"
+	"path"
 
 	"github.com/gluster/glusterd2/glusterd2/daemon"
 	"github.com/gluster/glusterd2/glusterd2/volume"
 	"github.com/gluster/glusterd2/glusterd2/xlator"
 	gderrors "github.com/gluster/glusterd2/pkg/errors"
 
 	log "github.com/sirupsen/logrus"
+	config "github.com/spf13/viper"
 )
 
 var names = [...]string{"replicate", "afr"}
 
 const (
-	selfHealKey = "self-heal-daemon"
+	selfHealKey          = "self-heal-daemon"
+	granularEntryHealKey = "granular-entry-heal"
 )
 
 type shdActor struct{}
 
 func (actor *shdActor) Do(v *volume.Volinfo, key string, value string, logger log.FieldLogger) error {
-	if key != selfHealKey {
+	if key != selfHealKey && key != granularEntryHealKey {
 		return nil
 	}
-	glustershDaemon, err := newGlustershd()
-	if err != nil {
-		return err
-	}
+	switch key {
+	case selfHealKey:
+		glustershDaemon, err := newGlustershd()
+		if err != nil {
+			return err
+		}
 
-	switch value {
-	case "on":
-		if isHealEnabled(v) {
-			err = daemon.Start(glustershDaemon, true, logger)
-			if err != nil && err != gderrors.ErrProcessAlreadyRunning {
-				return err
+		switch value {
+		case "on":
+			if isHealEnabled(v) {
+				err = daemon.Start(glustershDaemon, true, logger)
+				if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+					return err
+				}
+			}
+		case "off":
+			if !isHealEnabled(v) {
+				isVolRunning, err := volume.AreReplicateVolumesRunning()
+				if err != nil {
+					return err
+				}
+				if !isVolRunning {
+					return daemon.Stop(glustershDaemon, true, logger)
+				}
 			}
 		}
-	case "off":
-		if !isHealEnabled(v) {
-			isVolRunning, err := volume.AreReplicateVolumesRunning()
+	case granularEntryHealKey:
+		switch value {
+		case "enable":
```

Comment:
question:

though am not aware of the functionality, I have some questions
do we have ```disable``` value also for ```granularEntryHealKey```?
what happens to this when we stop or delete volume? do we need to stop or update ```glfsheal``` or not?

don't we need to add some logic in UNDO function to revert back the changes if Do() fails?

Reply:
@Madhu-1 In case of granular-entry-heal only enable needs an action which includes running glfsheal binary with some arguments just to check if heals are  not pending. In that case enable is allowed to happen else enable fails if there is any pending heals. In case of disable, there is no need of any action, only the xlator option needs to be set  to disable. 

When we stop or delete volume I don't think there is anything additional needed to be done, granular-entry-heal is purely shd's concern and is lookedup only when running heal process.

I don't think we can add anything in the UNDO function, because if the operation fails only xlator option needs to be restored.